File: leetcode/1921/solution.py
"""
https://leetcode.com/problems/eliminate-maximum-number-of-monsters/
1921. Eliminate Maximum Number of Monsters
"""

class Solution:
    def eliminateMaximum(self, dist: List[int], speed: List[int]) -> int:
        arr = [d // s + (1 if d % s else 0) for d, s in zip(dist, speed)]
        arr.sort()
        for mins_taken, mins_to_reach in enumerate(arr):
            if mins_taken >= mins_to_reach:
                return mins_taken
        return len(dist)

File: leetcode/1921/solution.rs
/**
 * https://leetcode.com/problems/eliminate-maximum-number-of-monsters/
 * 1921. Eliminate Maximum Number of Monsters
 */
impl Solution {
    pub fn eliminate_maximum(dist: Vec<i32>, speed: Vec<i32>) -> i32 {
        let mut minutes = dist
            .iter()
            .zip(speed.iter())
            .map(|(d, s)| d / s + if d % s > 0 {1} else {0})
            .collect::<Vec<i32>>();

        minutes.sort_unstable();
        for (i, &n) in minutes.iter().enumerate() {
            let i = i as i32;
            if i >= n {
                return i;
            }
        }
        
        // killed all monsters
        dist.len() as i32
    }
